include/elements/2d/triangle_lme_element.tcc

```diff
@@ -0,0 +1,423 @@
+//! Assign nodal connectivity property for LME elements
+template <unsigned Tdim>
+void mpm::TriangleLMEElement<Tdim>::initialise_lme_connectivity_properties(
+    double beta, double radius, bool anisotropy,
+    const Eigen::MatrixXd& nodal_coordinates) {
+  this->nconnectivity_ = nodal_coordinates.rows();
+  this->nodal_coordinates_ = nodal_coordinates;
+  this->beta_ = beta;
+  this->anisotropy_ = anisotropy;
+  this->support_radius_ = radius;
+}
+
+//! Return shape functions of a Quadrilateral LME Element at a given
+//! local coordinate
+template <unsigned Tdim>
+inline Eigen::VectorXd mpm::TriangleLMEElement<Tdim>::shapefn(
+    const Eigen::Matrix<double, Tdim, 1>& xi,
+    Eigen::Matrix<double, Tdim, 1>& lambda,
+    const Eigen::Matrix<double, Tdim, Tdim>& deformation_gradient) const {
+
+  //! To store shape functions
+  Eigen::VectorXd shapefn =
+      Eigen::VectorXd::Constant(this->nconnectivity_, 1.0);
+
+  //! Substitute the TriLME by the FEM shape funtion in the limit of the simplex
+  if (this->nconnectivity_ == 3)
+    return mpm::TriangleElement<Tdim, 3>::shapefn(xi, lambda,
+                                                  deformation_gradient);
+
+  try {
+    //! Convert local coordinates to real coordinates
+    Eigen::Matrix<double, Tdim, 1> pcoord =
+        Eigen::Matrix<double, Tdim, 1>::Zero();
+    auto local_shapefn = this->shapefn_local(xi, lambda, deformation_gradient);
+    for (unsigned i = 0; i < local_shapefn.size(); ++i)
+      pcoord.noalias() +=
+          local_shapefn(i) * nodal_coordinates_.row(i).transpose();
+
+    //! Create relative coordinate vector
+    const auto& rel_coordinates =
+        (-nodal_coordinates_.transpose()).colwise() + pcoord;
+
+    //! Create metric tensor
+    Eigen::Matrix<double, Tdim, Tdim> metric =
+        Eigen::Matrix<double, Tdim, Tdim>::Identity();
+
+    if (anisotropy_) {
+      //! Anisotropic metric tensor
+      const auto& inverse_deformation_gradient = deformation_gradient.inverse();
+      metric = inverse_deformation_gradient.transpose() *
+               inverse_deformation_gradient;
+    }
+
+    //! Compute particle-node distance once as a vector
+    Eigen::VectorXd distance =
+        Eigen::VectorXd::Constant(this->nconnectivity_, 0.0);
+    for (unsigned n = 0; n < this->nconnectivity_; ++n) {
+      distance(n) = std::sqrt((rel_coordinates.col(n)).transpose() *
+                              (metric * rel_coordinates.col(n)));
+    }
+
+    //! Compute functional f in each connectivity
+    Eigen::VectorXd f = Eigen::VectorXd::Constant(this->nconnectivity_, 0.0);
```

Comment:
I think that will complicate the call sequence. I prefer to keep it this way since it is quite readable right now. Besides, I dont think we can test local functions outside of the class.